core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java

```diff
@@ -266,14 +267,23 @@ private boolean usePendingAcquisition(SessionProtocol desiredProtocol, PoolKey k
             return false;
         }
 
+        numPendingAcquisitions++;
+
         timingsBuilder.pendingAcquisitionStart();
         pendingAcquisition.handle((pch, cause) -> {
             timingsBuilder.pendingAcquisitionEnd();
 
+            numPendingAcquisitions--;
+
             if (cause == null) {
                 final SessionProtocol actualProtocol = pch.protocol();
                 if (actualProtocol.isMultiplex()) {
-                    promise.complete(pch);
+                    final HttpSession session = HttpSession.get(pch.get());
+                    if (numPendingAcquisitions < session.maxUnfinishedResponses() - 1) {
```

Comment:
should also compare with `unfinishedResponses`.

also, technically this may not be a correct check -- should also consider timing with
1) adding response
2) check for acquireNowExact